### ISSUE TITLE
Escape key doesn't work (firefox)

### ISSUE THREAD

**Issue by j-cr** [philc/vimium#3699]
The extension doesn't react to the escape key in any context, thus making it impossible to exit to normal mode or to cancel an action.

- press f, then press escape
- expected: hints are removed, the page goes back to normal mode
- actual: nothing happens, the page stays in the link selection mode

Same with visual selection, vomnibar, etc; escape key doesn't work in any context.

Note that the browser only receives the key UP event for the escape key, not the key down event. I wonder if that has something to do with it. I've tested it in tridactyl too, same behavior. Any ideas?

Mozilla Firefox 68.9.0esr
Vimium 1.66

**Comment by j-cr** [comment]
Works as inteded with a fresh profile. The problem is profile-specific. Still keeping it open in case someone encountered a similar problem.

**Comment by gdh1995** [comment]
Could you do a simple test to find what Vimium got?
* open https://gdh1995.cn/vimium-c/keyboard-test.html
* click Vimium's icon and save options to make Vimium disabled on the page
* reload the page, and enable "disable Vimium / Vimium C / VimFx"
* then press Escape to see what it will log.

**Comment by j-cr** [comment]
@gdh1995 keyup only for escape, whether the extension is disabled or not. It makes no difference if it's enabled or disabled in regards to the escape key events.

As a side note, I noticed sometimes I get key up events from normal navigation keys (j, k, etc) too when I press them in a quick succession. Unrelated, but curious maybe.

**Comment by gdh1995** [comment]
Then this is an issue of Firefox itself, or does another extension hook the keydown events?

**Comment by j-cr** [comment]
I've only been able to reproduce the issue with firefox and with a single specific profile that I use. In this profile, neither vimium nor tridactyl receive the escape key. On a fresh profile however both extensions work fine.

Of course, I've tried disabling every other extension except for vimium, but that doesn't affect the issue; so I can't pinpoint the cause, and I'm not sure what else I can try to find the cause. Some kind of about:config switch that affects escape key behavior?.. I'm running out of ideas.

Is it possible to remap (via config) at least some of the "unfocus"/"return to normal mode" commands to another key? E.g. make it so pressing space exits the link selection mode? Or make C-[ do the same, as it does in vim?

**Comment by gdh1995** [comment]
You may use `<c-[>`, and it works like `<esc>` at most places (when `useVimLikeEscape` is true).

And `KeyboardUtils.isEscape` supports `mapkey <anykey> <c-[>` in key mappings.

**Comment by j-cr** [comment]
Oh, `C-[` is supposed to work out of the box? It doesn't work for me. On https://gdh1995.cn/vimium-c/keyboard-test.html I see that the key down events gets "eaten" by vimium when it's enabled, but it produces absolutely no effect.

The mapkey trick however works! `mapkey <c-g> <c-[>` makes C-g work like escape in all relevant contexts. I find it to be quite weird, since C-[ itself doesn't work at all. Any ideas about what's happening?

Also, I noticed that as a workaround I can use backspace in link selection mode and find mode (but not in vomnibar).

**Comment by gdh1995** [comment]
What will the test page log for `Ctrl+[` on your Firefox when Vimium is disabled?

**Comment by j-cr** [comment]
keydown control
keyup [
keyup control

...well, which is exactly the same as when it's enabled :/ Same happens for C-], but for C-S-[ (aka `C-{`) or C-' or ctrl + any alphanumerics the behavior is normal (both key down and key up events are displayed for both ctrl and the second key). I'm quite confused, it seems like firefox specifically treats escape, ctrl+[ and ctrl+] in a weird way.

**Comment by trysten** [comment]
Just noticed this is happening to me when trying to escape from github's finder feature. Vimium is eating the `esc` somehow: the problem resolves when using a blank rule to disable Vimium. Keycodes are as expected for `esc` and `C-[`

**Comment by arisolt** [comment]
I have the exact same issue. The keydown event for the Escape key isn't being registered on e.g https://w3c.github.io/uievents/tools/key-event-viewer.html. Troubleshoot mode (safe mode = all extensions disabled) doesn't fix it, but creating a new profile does. I also suspected something in about:config, but nothing stood out. Very strange!

**Comment by FriederRodewald** [comment]
Had the issue that <esc> was not working and only <c-esc> exited modes as expected. Problem was solved for me after resetting "permissions.default.shortcuts" in about:config, which I have seem to changed a while back :)

**Comment by arisolt** [comment]
> Had the issue that 'esc' was not working and only 'c-esc' exited modes as expected. Problem was solved for me after resetting "permissions.default.shortcuts" in about:config, which I have seem to changed a while back :) Maybe this helps someone in the future...

This fixed it for me too! Thank you!

Unfortunately this was a very useful setting, because it prevented websites like Twitter and Github from hijacking keys. But it shouldn't have interferred with the Esc keypress events. It could be a bug in Firefox.

**Comment by gdh1995** [comment]
During my tests on Firefox 122, Win11, `keydown#key=Escape` will be blocked by `permissions.default.shortcuts=2`, while `Ctrl+[` and `Ctrl+]` works as intended.

**Comment by adamency** [comment]
@gdh1995 is absolutely right. I had the same issue for a long time, and this is exactly what the problem was `permissions.default.shortcuts` set to `2` is `about:config`. And for reference, it doesn't just **block the Escape** key for Vimium, but for **ALL Firefox extensions** (i.e Ublock zapper/picker for instance).

We need to contact Mozilla about this as this is a huge issue for keyboard-centric users on Firefox.

**Comment by blackxored** [comment]
Was this fixed, I'm still seeing the issue on Firefox 128.

**Comment by philg-dev** [comment]
@blackxored are you using `permissions.default.shortcuts = 2`? Resetting that to default should probably solve your issue if I read the comments above correctly.

Here's a reference to "Betterfox" with some explanation of the setting:
https://github.com/yokoffing/Betterfox/blob/429abb61f52726b7c6f1675bccca32bbb5359c02/Peskyfox.js#L531

**Comment by blackxored** [comment]
Yes I'm using that setting, I was hoping there's a way to whitelist Vimium somehow?

**Comment by philg-dev** [comment]
> Yes I'm using that setting, I was hoping there's a way to whitelist Vimium somehow?
> 
> EDIT: It's gonna be kind of hard to enable, since Schizofox does this by default without a toggle: https://github.com/schizofox/schizofox/blob/0c2c713405fb02693992c823d98f43a36efed7a5/flake/modules/home-manager/firefox/preferences/default.nix#L490

I'm not super familiar with NIX and HomeManager, but from what I know... Shouldn't it be possible to just add the property to your HomeManager configuration for that schizofox module or whatever it is?! Or literally just change the property in the `*.nix` file.

I don't think there's a way to whitelist Vimium as long as `permissions.default.shortcuts = 2` is active.
